### DISCUSSION TITLE
Welcome to bitser Discussions!

### DISCUSSION THREAD

**Discussion by gvx** [gvx/bitser#19] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by WhyNotDogie** [comment]
i like this library, but bitser is uhm... not a dutch word. at least, ive never heard it in my many years in holland

**Reply by gvx** [reply]
I'm happy that you like it! "Bitser" is the comparative (vergrotende trap) of "bits", which means abrupt, snappy, unfriendly.

https://www.vandale.nl/gratis-woordenboek/nederlands/betekenis/bitser

**Comment by masakk1** [comment]
I'm a little late to the party, but I would like to get some tutorials regarding class registration.
For example, I was wondering if I could use it to make my packet size smaller, but binser is so dead that not even their examples work.

Here is my idea:
```lua
--#region Packet types
---Every type of packet.
---@class PKT_TYPE
PKT_TYPE = {
	CONNECT = {
		REQUEST = 11,
		ACCEPT = 12,
	},
}

-- templates of how different packets would look
PKT_TEMPLATE = {}
PKT_TEMPLATE.CLIENT = {
	"type",
	"clientID",
	data = {},
}
PKT_TEMPLATE.SERVER = {
	"type",
	data = {},
}
--#endregion
-- Then it'd be great if I could "register" these tables 
bitser.register("ClientBasePacket", PKT_TEMPLATE.CLIENT)

-- then, creating a packet from a client
function createClientPacket(type, data)
    local pkt = {
        type = type,
        data = data or nil,
    }
    local serialised = biter.dumpclass("ClientBasePacket", pkt)
    return serialised
end
```
This obviously doesn't work right now, but it's how I *think* should work. I don't have any documentation to check for how to do it, and bitser changes a few things from binser, and even binser's documentation is sparse or just wrong.

I could also explore the idea of static byte lengths for the payload's headers, but I haven't gotten explored that yet, for now.

**Reply by masakk1** [reply]
I wouldn't mind writing documentation for it. Once I understood it, of course.

**Reply by gvx** [reply]
There is a tentative implementation of something like this in the [examples directory](https://github.com/gvx/bitser/tree/master/examples) -- it doesn't support templates with nested tables yet for example.

It does not work on plain tables like in your example, however, and I have no intention of making this work, because this would mean that you couldn't e.g. serialize a sequence of packets, you'd need to preprocess each of them individually, and at that point you may just as well just do something like:

```lua
function createClientPacket(type, clientID, data)
    return bitser.dumps {type, clientID, data}
end
```

Or alternatively, if you use classes, you could do something like:

```lua
local bitser = require 'bitser'
local bitser_templates = require 'bitser-templates'
local class = require 'your-favourite-class-library'

ClientBasePacket = class('ClientBasePacket')

ClientBasePacket._template = {
	"type",
	"clientID",
	data = {},
}

bitser_templates.registerClass(ClientBasePacket)

-- then, creating a packet from a client
function createClientPacket(pkt) -- where pkt is an instance of ClientBasePacket
    return bitser.dumps(pkt)
end
```
